src/implementations/astroport/helpers.rs

```diff
@@ -78,3 +83,112 @@ pub(crate) fn astro_asset_info_to_cw_asset_info(asset_info: &AstroAssetInfo) ->
+        } else {
+            let amp_range = init_amp - next_amp;
+            let res = init_amp - (amp_range * elapsed_time).checked_div(time_range)?;
+            Ok(res.u128() as u64)
+        }
+    } else {
+        Ok(config.next_amp)
+    }
+}
+
+/// ## Description
+/// Computes the stableswap invariant (D).
+///
+/// * **Equation**
+///
+/// A * sum(x_i) * n**n + D = A * D * n**n + D**(n+1) / (n**n * prod(x_i))
+///
+/// ## Params
+/// * **leverage** is an object of type [`u128`].
+///
+/// * **amount_a** is an object of type [`u128`].
+///
+/// * **amount_b** is an object of type [`u128`].
+pub fn compute_d(leverage: u64, amount_a: u128, amount_b: u128) -> Option<u128> {
+    let amount_a_times_coins =
+        checked_u8_mul(&U256::from(amount_a), N_COINS)?.checked_add(U256::one())?;
+    let amount_b_times_coins =
+        checked_u8_mul(&U256::from(amount_b), N_COINS)?.checked_add(U256::one())?;
+    let sum_x = amount_a.checked_add(amount_b)?; // sum(x_i), a.k.a S
+    if sum_x == 0 {
+        Some(0)
+    } else {
+        let mut d_previous: U256;
+        let mut d: U256 = sum_x.into();
+
+        // Newton's method to approximate D
+        for _ in 0..ITERATIONS {
+            let mut d_product = d;
+            d_product = d_product
+                .checked_mul(d)?
+                .checked_div(amount_a_times_coins)?;
+            d_product = d_product
+                .checked_mul(d)?
+                .checked_div(amount_b_times_coins)?;
+            d_previous = d;
+            // d = (leverage * sum_x + d_p * n_coins) * d / ((leverage - 1) * d + (n_coins + 1) * d_p);
+            d = calculate_step(&d, leverage, sum_x, &d_product)?;
+            // Equality with the precision of 1
+            if d == d_previous {
+                break;
+            }
+        }
+        u128::try_from(d).ok()
+    }
+}
+
+/// ## Description
+/// Returns self multiplied by b.
+fn checked_u8_mul(a: &U256, b: u8) -> Option<U256> {
+    let mut result = *a;
+    for _ in 1..b {
+        result = result.checked_add(*a)?;
+    }
+    Some(result)
+}
+
+/// ## Description
+/// Helper function used to calculate the D invariant as a last step in the `compute_d` public function.
+///
+/// * **Equation**:
+///
+/// d = (leverage * sum_x + d_product * n_coins) * initial_d / ((leverage - 1) * initial_d + (n_coins + 1) * d_product)
+fn calculate_step(initial_d: &U256, leverage: u64, sum_x: u128, d_product: &U256) -> Option<U256> {
+    let leverage_mul = U256::from(leverage).checked_mul(sum_x.into())? / AMP_PRECISION;
+    let d_p_mul = checked_u8_mul(d_product, N_COINS)?;
+
+    let l_val = leverage_mul.checked_add(d_p_mul)?.checked_mul(*initial_d)?;
+
+    let leverage_sub =
+        initial_d.checked_mul((leverage.checked_sub(AMP_PRECISION)?).into())? / AMP_PRECISION;
+    let n_coins_sum = checked_u8_mul(d_product, N_COINS.checked_add(1)?)?;
+
+    let r_val = leverage_sub.checked_add(n_coins_sum)?;
+
+    l_val.checked_div(r_val)
+}
```

Comment:
could you link to the original astroport implementation of these functions on the doc comments